File: ALEO/collect_aleo_context.mjs
import fs from "node:fs/promises";
import path from "node:path";

const ROOT = process.cwd();
const OUT = path.join(ROOT, "aleo_context");
const API = "https://api.explorer.provable.com/v2";
const API_V1 = "https://api.explorer.provable.com/v1";
const NETWORKS = ["mainnet", "testnet"];

async function getJson(url) {
  const response = await fetch(url, { headers: { "user-agent": "aleo-context-collector/1.0" } });
  if (!response.ok) throw new Error(`${response.status} ${response.statusText}: ${await response.text()}`);
  return response.json();
}

async function getText(url) {
  const value = await getJson(url);
  return typeof value === "string" ? value : JSON.stringify(value, null, 2);
}

async function writeText(filePath, content) {
  await fs.mkdir(path.dirname(filePath), { recursive: true });
  await fs.writeFile(filePath, content, "utf8");
}

function safeName(programId) {
  return programId.replaceAll("/", "_");
}

function unique(values) {
  return [...new Set(values.filter(Boolean))].sort();
}

function matchAll(source, regex) {
  return [...source.matchAll(regex)];
}

function parseBlockItems(source, keyword) {
  const heading = new RegExp(`^${keyword}\\s+([A-Za-z0-9_]+):\\s*$`, "gm");
  const matches = matchAll(source, heading);
  const items = {};
  for (let i = 0; i < matches.length; i += 1) {
    const name = matches[i][1];
    const start = matches[i].index + matches[i][0].length;
    let end = i + 1 < matches.length ? matches[i + 1].index : source.length;
    const tail = source.slice(start);
    const nextTop = tail.search(/^(mapping|struct|record|function|finalize|closure|constructor)\s+[A-Za-z0-9_]+:|^import\s+/m);
    if (nextTop >= 0) end = Math.min(end, start + nextTop);
    items[name] = source.slice(start, end).replace(/^\n+|\n+$/g, "");
  }
  return items;
}

function parseIo(body = "") {
  const inputs = matchAll(body, /^\s*input\s+([A-Za-z0-9_]+)\s+as\s+([^;]+);/gm).map((m) => ({
    register: m[1],
    type: m[2].trim(),
  }));
  const outputs = matchAll(body, /^\s*output\s+([A-Za-z0-9_]+)\s+as\s+([^;]+);/gm).map((m) => ({
    register: m[1],
    type: m[2].trim(),
  }));
  return { inputs, outputs };
}

function parseMapping(body = "") {
  return {
    key: body.match(/key\s+as\s+([^;]+);/)?.[1]?.trim() ?? null,
    value: body.match(/value\s+as\s+([^;]+);/)?.[1]?.trim() ?? null,
  };
}

function parseFields(body = "") {
  return matchAll(body, /^\s*([A-Za-z0-9_]+)\s+as\s+([^;]+);/gm).map((m) => ({
    name: m[1],
    type: m[2].trim(),
  }));
}

function refsIn(text, programId) {
  return unique(matchAll(text ?? "", /\b([a-z][a-z0-9_]*\.aleo)(?=\/|::)/g).map((m) => m[1]).filter((x) => x !== programId));
}

function stateWrites(text = "") {
  const setWrites = matchAll(text, /\bset\s+[^;\n]+?\s+into\s+([A-Za-z0-9_]+)\[/g).map((m) => m[1]);
  const removes = matchAll(text, /\bremove\s+([A-Za-z0-9_]+)\[/g).map((m) => m[1]);
  return unique([...setWrites, ...removes]);
}

function conditionMarkers(text = "") {
  return unique(matchAll(text, /\b(assert\.[a-z]+|branch\.[a-z]+|ternary|gte|lte|lt|gt|is\.eq|nor|and|or)\b/g).map((m) => m[1]));
}

function parseProgram(programId, source, calls) {
  const functions = parseBlockItems(source, "function");
  const finalizers = parseBlockItems(source, "finalize");
  const mappings = parseBlockItems(source, "mapping");
  const structs = parseBlockItems(source, "struct");
  const records = parseBlockItems(source, "record");
  const imports = unique(matchAll(source, /^import\s+([a-z][a-z0-9_]*\.aleo);/gm).map((m) => m[1]));
  const parsedFunctions = {};

  for (const [name, body] of Object.entries(functions)) {
    const finalize = finalizers[name] ?? "";
    const { inputs, outputs } = parseIo(body);
    parsedFunctions[name] = {
      inputs,
      outputs,
      has_finalize: Boolean(finalizers[name]),
      async: body.includes("async "),
      returns_future: body.includes(".future"),
      cross_contract_refs: refsIn(body, programId),
      state_reads: unique(matchAll(body, /\b(?:get|get\.or_use|contains)\s+([A-Za-z0-9_]+)\[/g).map((m) => m[1])),
      state_writes: stateWrites(body),
      conditional_markers: conditionMarkers(body),
      finalize_inputs: parseIo(finalize).inputs,
      finalize_state_reads: unique(matchAll(finalize, /\b(?:get|get\.or_use|contains)\s+([A-Za-z0-9_]+)\[/g).map((m) => m[1])),
      finalize_state_writes: stateWrites(finalize),
      finalize_uses_block_height: finalize.includes("block.height"),
      finalize_conditional_markers: conditionMarkers(finalize),
      snark_verify: body.includes("snark.verify") || finalize.includes("snark.verify"),
    };
  }

  const asyncOrConditional = Object.entries(parsedFunctions)
    .filter(([, data]) => data.async || data.has_finalize || data.finalize_uses_block_height || data.conditional_markers.length || data.finalize_conditional_markers.length)
    .map(([name]) => name);

  return {
    program_id: programId,
    calls_24h_or_metric_window: calls,
    imports,
    mappings: Object.fromEntries(Object.entries(mappings).map(([name, body]) => [name, parseMapping(body)])),
    structs: Object.fromEntries(Object.entries(structs).map(([name, body]) => [name, parseFields(body)])),
    records: Object.fromEntries(Object.entries(records).map(([name, body]) => [name, parseFields(body)])),
    functions: parsedFunctions,
    finalizers: Object.keys(finalizers).sort(),
    cross_contract_refs: refsIn(source, programId),
    async_or_conditional_paths: asyncOrConditional,
    state_commitment_surfaces: {
      records: Object.keys(records).sort(),
      mappings: Object.keys(mappings).sort(),
      future_outputs: Object.entries(parsedFunctions).flatMap(([fn, data]) =>
        data.outputs.filter((output) => output.type.endsWith(".future")).map((output) => ({ function: fn, output: output.type })),
      ),
    },
    event_model_note:
      "Aleo does not expose EVM-style events in program source; observable commitments are records, transitions, futures/finalize effects, and mapping updates.",
  };
}

function classifyPath(programId) {
  const p = programId.toLowerCase();
  if (/(bridge|warp|hyp_|mailbox|gmp|connection|vlink)/.test(p)) return "bridge/cross-chain";
  if (/(stake|staking|delegator|validator|pondo|whalepool)/.test(p)) return "staking/liquid staking";
  if (/(token|stablecoin|usdc|usad|wrapped|credits)/.test(p)) return "token/credit flow";
  if (/(oracle|checksum|rate_limit)/.test(p)) return "oracle/governance guard";
  if (/(nft|ticket|arcade|mint)/.test(p)) return "NFT/game asset";
  if (/(pool|swap|lend|flash|arcn|dara)/.test(p)) return "DeFi pool/lending";
  return "application/other";
}

async function collectNetwork(network) {
  const rawDir = path.join(OUT, "raw", network);
  const sourceDir = path.join(rawDir, "program_sources");
  const parsedDir = path.join(OUT, "parsed", network);
  const metrics = await getJson(`${API}/${network}/metrics/programs`);
  await writeText(path.join(rawDir, "program_metrics.json"), JSON.stringify(metrics, null, 2));
  await writeText(path.join(rawDir, "latest_height.txt"), `${(await getText(`${API_V1}/${network}/latest/height`)).trim()}\n`);

  const parsed = [];
  const failures = [];
  for (const item of metrics) {
    try {
      const source = await getText(`${API}/${network}/program/${item.program_id}`);
      await writeText(path.join(sourceDir, `${safeName(item.program_id)}.aleo`), source);
      const data = parseProgram(item.program_id, source, item.calls ?? 0);
      await writeText(path.join(parsedDir, `${safeName(item.program_id)}.json`), JSON.stringify(data, null, 2));
      parsed.push(data);
      await new Promise((resolve) => setTimeout(resolve, 80));
    } catch (error) {
      failures.push({ program_id: item.program_id, error: String(error.message ?? error) });
    }
  }

  await writeText(path.join(rawDir, "download_failures.json"), JSON.stringify(failures, null, 2));
  await writeText(path.join(parsedDir, "programs_summary.json"), JSON.stringify(parsed, null, 2));
  const csv = [
    "program_id,calls,functions,mappings,records,imports,cross_contract_refs,async_or_conditional_paths",
    ...parsed.map((p) =>
      [
        p.program_id,
        p.calls_24h_or_metric_window,
        Object.keys(p.functions).length,
        Object.keys(p.mappings).length,
        Object.keys(p.records).length,
        `"${p.imports.join(" ")}"`,
        `"${p.cross_contract_refs.join(" ")}"`,
        `"${p.async_or_conditional_paths.join(" ")}"`,
      ].join(","),
    ),
  ].join("\n");
  await writeText(path.join(parsedDir, "programs_summary.csv"), `${csv}\n`);
  return { parsed, failures };
}

async function writeGraphs(allData) {
  const dot = ["digraph aleo_program_dependencies {", "  rankdir=LR;"];
  const md = ["# Aleo dependency graph", ""];
  for (const [network, programs] of Object.entries(allData)) {
    md.push(`## ${network}`);
    for (const program of programs) {
      const refs = unique([...program.imports, ...program.cross_contract_refs]);
      if (refs.length) {
        for (const ref of refs) dot.push(`  "${network}:${program.program_id}" -> "${network}:${ref}";`);
        md.push(`- \`${program.program_id}\` -> ${refs.map((ref) => `\`${ref}\``).join(", ")}`);
      } else {
        dot.push(`  "${network}:${program.program_id}";`);
      }
    }
    md.push("");
  }
  dot.push("}");
  await writeText(path.join(OUT, "graphs", "dependencies.dot"), `${dot.join("\n")}\n`);
  await writeText(path.join(OUT, "graphs", "dependencies.md"), md.join("\n"));
}

async function writeReports(allData, allFailures) {
  const lines = [
    "# Aleo context map",
    "",
    `Generated at: ${new Date().toISOString()}`,
    "",
    "Scope: active programs returned by Provable `/v2/{network}/metrics/programs`, with full source pulled from `/v2/{network}/program/{program_id}`.",
    "AleoScan registry pages were used as a secondary explorer reference, but local collection avoids Cloudflare-protected scraping.",
    "",
    "## Source folders",
    "",
    "- `raw/mainnet/program_sources/` and `raw/testnet/program_sources/`: downloaded Aleo Instruction sources.",
    "- `parsed/{network}/`: per-program JSON plus CSV summaries.",
    "- `graphs/dependencies.dot`: cross-program reference graph.",
    "- `reports/high_risk_async_paths.md`: async, delayed, conditional, and bridge/proof-sensitive paths.",
    "",
  ];

  for (const [network, programs] of Object.entries(allData)) {
    const totalCalls = programs.reduce((sum, p) => sum + p.calls_24h_or_metric_window, 0);
    lines.push(`## ${network}`, "", `- Programs collected: ${programs.length}`, `- Metric-window calls represented: ${totalCalls}`, `- Download failures: ${allFailures[network].length}`, "");
    lines.push("| Program | Calls | Class | Functions | Mappings | Cross refs | Async/conditional paths |");
    lines.push("|---|---:|---|---:|---:|---|---:|");
    for (const p of [...programs].sort((a, b) => b.calls_24h_or_metric_window - a.calls_24h_or_metric_window)) {
      lines.push(
        `| \`${p.program_id}\` | ${p.calls_24h_or_metric_window} | ${classifyPath(p.program_id)} | ${Object.keys(p.functions).length} | ${Object.keys(p.mappings).length} | ${p.cross_contract_refs.slice(0, 6).join(", ") || "-"} | ${p.async_or_conditional_paths.length} |`,
      );
    }
    lines.push("");
  }
  await writeText(path.join(OUT, "reports", "context_map.md"), lines.join("\n"));

  const risky = [
    "# Async, delayed, conditional, bridge/proof-sensitive paths",
    "",
    "Selection criteria: `async`, `finalize`, `.future`, `block.height`, branches/assertions/ternaries, `snark.verify`, or names indicating bridge/oracle/staking/token economic paths.",
    "",
  ];
  for (const [network, programs] of Object.entries(allData)) {
    risky.push(`## ${network}`);
    for (const p of [...programs].sort((a, b) => b.calls_24h_or_metric_window - a.calls_24h_or_metric_window)) {
      const importantName = classifyPath(p.program_id) !== "application/other";
      if (!p.async_or_conditional_paths.length && !importantName) continue;
      risky.push(`### \`${p.program_id}\` (${p.calls_24h_or_metric_window} calls, ${classifyPath(p.program_id)})`);
      if (p.cross_contract_refs.length) risky.push(`Cross-contract refs: ${p.cross_contract_refs.map((x) => `\`${x}\``).join(", ")}`);
      if (Object.keys(p.mappings).length) risky.push(`State mappings: ${Object.keys(p.mappings).map((x) => `\`${x}\``).join(", ")}`);
      for (const fn of p.async_or_conditional_paths) {
        const data = p.functions[fn];
        const flags = [];
        if (data.async) flags.push("async");
        if (data.has_finalize) flags.push("finalize");
        if (data.finalize_uses_block_height) flags.push("block.height delay");
        if (data.snark_verify) flags.push("snark.verify");
        if (data.cross_contract_refs.length) flags.push("cross-contract");
        risky.push(`- \`${fn}\`: ${flags.join(", ") || "conditional"}`);
      }
      risky.push("");
    }
  }
  await writeText(path.join(OUT, "reports", "high_risk_async_paths.md"), risky.join("\n"));

  const econ = [
    "# Economic paths and activity",
    "",
    "Calls are from Provable program metrics at collection time. TVL is not recomputed from private records; use this as a routing/activity map, not a balance attestation.",
    "",
  ];
  for (const [network, programs] of Object.entries(allData)) {
    econ.push(`## ${network}`);
    for (const p of [...programs].sort((a, b) => b.calls_24h_or_metric_window - a.calls_24h_or_metric_window)) {
      const cls = classifyPath(p.program_id);
      if (cls === "application/other" && p.calls_24h_or_metric_window < 50) continue;
      econ.push(`- \`${p.program_id}\`: ${p.calls_24h_or_metric_window} calls; class \`${cls}\`; state \`${Object.keys(p.mappings).slice(0, 8).join(", ") || "no public mappings parsed"}\`.`);
    }
    econ.push("");
  }
  await writeText(path.join(OUT, "reports", "economic_paths.md"), econ.join("\n"));
}

async function main() {
  await fs.mkdir(OUT, { recursive: true });
  const allData = {};
  const allFailures = {};
  for (const network of NETWORKS) {
    const { parsed, failures } = await collectNetwork(network);
    allData[network] = parsed;
    allFailures[network] = failures;
  }
  await writeGraphs(allData);
  await writeReports(allData, allFailures);
  await writeText(
    path.join(OUT, "sources.json"),
    JSON.stringify(
      {
        provable_v2_program_metrics: `${API}/{network}/metrics/programs`,
        provable_v2_program_source: `${API}/{network}/program/{program_id}`,
        provable_v2_program_mappings: `${API}/{network}/program/{program_id}/mappings`,
        provable_v1_latest_height: `${API_V1}/{network}/latest/height`,
        aleo_docs_programs: "https://developer.aleo.org/apis/v2/programs/",
        aleo_docs_program_metrics: "https://developer.aleo.org/apis/v2/program-metrics/",
        aleo_docs_defi: "https://developer.aleo.org/apis/v2/de-fi/",
        aleo_docs_staking: "https://developer.aleo.org/apis/v2/staking/",
        aleoscan_registry_reference: "https://aleoscan.io/programs and https://testnet.aleoscan.io/programs",
      },
      null,
      2,
    ),
  );
  console.log(`Wrote Aleo context to ${OUT}`);
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: ALEO/collect_aleo_context.py
#!/usr/bin/env python3
import csv
import json
import re
import time
import urllib.error
import urllib.request
from datetime import datetime, timezone
from pathlib import Path


ROOT = Path(__file__).resolve().parent
OUT = ROOT / "aleo_context"
API = "https://api.explorer.provable.com/v2"
API_V1 = "https://api.explorer.provable.com/v1"
NETWORKS = ("mainnet", "testnet")


def get_json(url):
    req = urllib.request.Request(url, headers={"User-Agent": "aleo-context-collector/1.0"})
    with urllib.request.urlopen(req, timeout=30) as response:
        raw = response.read().decode("utf-8")
    return json.loads(raw)


def get_text(url):
    value = get_json(url)
    return value if isinstance(value, str) else json.dumps(value, indent=2)


def write_text(path, content):
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")


def safe_name(program_id):
    return program_id.replace("/", "_")


def parse_block_items(source, keyword):
    items = {}
    pattern = re.compile(rf"^{keyword}\s+([A-Za-z0-9_]+):\s*$", re.MULTILINE)
    matches = list(pattern.finditer(source))
    for index, match in enumerate(matches):
        name = match.group(1)
        start = match.end()
        end = matches[index + 1].start() if index + 1 < len(matches) else len(source)
        next_top = re.search(
            r"^(mapping|struct|record|function|finalize|closure|constructor)\s+[A-Za-z0-9_]+:|^import\s+",
            source[start:],
            re.MULTILINE,
        )
        if next_top:
            end = min(end, start + next_top.start())
        body = source[start:end].strip("\n")
        items[name] = body
    return items


def parse_io(body):
    inputs = re.findall(r"^\s*input\s+([A-Za-z0-9_]+)\s+as\s+([^;]+);", body, re.MULTILINE)
    outputs = re.findall(r"^\s*output\s+([A-Za-z0-9_]+)\s+as\s+([^;]+);", body, re.MULTILINE)
    return [{"register": a, "type": b.strip()} for a, b in inputs], [
        {"register": a, "type": b.strip()} for a, b in outputs
    ]


def parse_mapping(body):
    key = re.search(r"key\s+as\s+([^;]+);", body)
    value = re.search(r"value\s+as\s+([^;]+);", body)
    return {
        "key": key.group(1).strip() if key else None,
        "value": value.group(1).strip() if value else None,
    }


def parse_fields(body):
    fields = []
    for name, typ in re.findall(r"^\s*([A-Za-z0-9_]+)\s+as\s+([^;]+);", body, re.MULTILINE):
        fields.append({"name": name, "type": typ.strip()})
    return fields


def parse_program(program_id, source, calls):
    functions = parse_block_items(source, "function")
    finalizers = parse_block_items(source, "finalize")
    mappings = parse_block_items(source, "mapping")
    structs = parse_block_items(source, "struct")
    records = parse_block_items(source, "record")

    parsed_functions = {}
    cross_refs = sorted(
        set(
            ref
            for ref in re.findall(r"\b([a-z][a-z0-9_]*\.aleo)(?=/|::)", source)
            if ref != program_id
        )
    )
    imports = re.findall(r"^import\s+([a-z][a-z0-9_]*\.aleo);", source, re.MULTILINE)

    for name, body in functions.items():
        inputs, outputs = parse_io(body)
        finalize = finalizers.get(name)
        parsed_functions[name] = {
            "inputs": inputs,
            "outputs": outputs,
            "has_finalize": finalize is not None,
            "async": "async " in body,
            "returns_future": ".future" in body,
            "cross_contract_refs": sorted(
                set(ref for ref in re.findall(r"\b([a-z][a-z0-9_]*\.aleo)(?=/|::)", body) if ref != program_id)
            ),
            "state_reads": sorted(set(re.findall(r"\b(?:get|get\.or_use|contains)\s+([A-Za-z0-9_]+)\[", body))),
            "state_writes": sorted(set(re.findall(r"\b(?:set|remove)\s+[^;\n]+?\s+into\s+([A-Za-z0-9_]+)\[|\bremove\s+([A-Za-z0-9_]+)\[", body))),
            "conditional_markers": sorted(set(re.findall(r"\b(assert\.[a-z]+|branch\.[a-z]+|ternary|gte|lte|lt|gt|is\.eq|nor|and|or)\b", body))),
            "finalize_inputs": parse_io(finalize)[0] if finalize else [],
            "finalize_state_reads": sorted(set(re.findall(r"\b(?:get|get\.or_use|contains)\s+([A-Za-z0-9_]+)\[", finalize or ""))),
            "finalize_state_writes": sorted(
                set(
                    item
                    for pair in re.findall(
                        r"\bset\s+[^;\n]+?\s+into\s+([A-Za-z0-9_]+)\[|\bremove\s+([A-Za-z0-9_]+)\[",
                        finalize or "",
                    )
                    for item in pair
                    if item
                )
            ),
            "finalize_uses_block_height": "block.height" in (finalize or ""),
            "finalize_conditional_markers": sorted(
                set(re.findall(r"\b(assert\.[a-z]+|branch\.[a-z]+|ternary|gte|lte|lt|gt|is\.eq|nor|and|or)\b", finalize or ""))
            ),
            "snark_verify": "snark.verify" in body or "snark.verify" in (finalize or ""),
        }

    async_paths = [
        name
        for name, data in parsed_functions.items()
        if data["async"]
        or data["has_finalize"]
        or data["finalize_uses_block_height"]
        or data["finalize_conditional_markers"]
        or data["conditional_markers"]
    ]

    return {
        "program_id": program_id,
        "calls_24h_or_metric_window": calls,
        "imports": sorted(set(imports)),
        "mappings": {name: parse_mapping(body) for name, body in mappings.items()},
        "structs": {name: parse_fields(body) for name, body in structs.items()},
        "records": {name: parse_fields(body) for name, body in records.items()},
        "functions": parsed_functions,
        "finalizers": sorted(finalizers.keys()),
        "cross_contract_refs": cross_refs,
        "async_or_conditional_paths": async_paths,
        "state_commitment_surfaces": {
            "records": sorted(records.keys()),
            "mappings": sorted(mappings.keys()),
            "future_outputs": [
                {"function": name, "output": output["type"]}
                for name, data in parsed_functions.items()
                for output in data["outputs"]
                if output["type"].endswith(".future")
            ],
        },
        "event_model_note": "Aleo does not expose EVM-style events in program source; observable commitments are records, transitions, futures/finalize effects, and mapping updates.",
    }


def collect_network(network):
    raw_dir = OUT / "raw" / network
    source_dir = raw_dir / "program_sources"
    parsed_dir = OUT / "parsed" / network

    metrics = get_json(f"{API}/{network}/metrics/programs")
    write_text(raw_dir / "program_metrics.json", json.dumps(metrics, indent=2))

    latest_height = get_text(f"{API_V1}/{network}/latest/height")
    write_text(raw_dir / "latest_height.txt", latest_height.strip() + "\n")

    parsed = []
    failures = []
    for item in metrics:
        program_id = item["program_id"]
        calls = item.get("calls", 0)
        try:
            source = get_text(f"{API}/{network}/program/{program_id}")
            write_text(source_dir / f"{safe_name(program_id)}.aleo", source)
            data = parse_program(program_id, source, calls)
            write_text(parsed_dir / f"{safe_name(program_id)}.json", json.dumps(data, indent=2))
            parsed.append(data)
            time.sleep(0.08)
        except (urllib.error.HTTPError, urllib.error.URLError, TimeoutError, json.JSONDecodeError) as exc:
            failures.append({"program_id": program_id, "error": str(exc)})

    write_text(raw_dir / "download_failures.json", json.dumps(failures, indent=2))
    write_text(parsed_dir / "programs_summary.json", json.dumps(parsed, indent=2))

    with (parsed_dir / "programs_summary.csv").open("w", newline="", encoding="utf-8") as handle:
        writer = csv.writer(handle)
        writer.writerow(
            [
                "program_id",
                "calls",
                "functions",
                "mappings",
                "records",
                "imports",
                "cross_contract_refs",
                "async_or_conditional_paths",
            ]
        )
        for data in parsed:
            writer.writerow(
                [
                    data["program_id"],
                    data["calls_24h_or_metric_window"],
                    len(data["functions"]),
                    len(data["mappings"]),
                    len(data["records"]),
                    " ".join(data["imports"]),
                    " ".join(data["cross_contract_refs"]),
                    " ".join(data["async_or_conditional_paths"]),
                ]
            )

    return parsed, failures


def write_graphs(all_data):
    graph_dir = OUT / "graphs"
    lines = ["digraph aleo_program_dependencies {", "  rankdir=LR;"]
    md = ["# Aleo dependency graph", ""]
    for network, programs in all_data.items():
        md.append(f"## {network}")
        for program in programs:
            refs = sorted(set(program["imports"] + program["cross_contract_refs"]))
            if refs:
                for ref in refs:
                    lines.append(f'  "{network}:{program["program_id"]}" -> "{network}:{ref}";')
                md.append(f"- `{program['program_id']}` -> " + ", ".join(f"`{ref}`" for ref in refs))
            else:
                lines.append(f'  "{network}:{program["program_id"]}";')
        md.append("")
    lines.append("}")
    write_text(graph_dir / "dependencies.dot", "\n".join(lines) + "\n")
    write_text(graph_dir / "dependencies.md", "\n".join(md))


def classify_path(program_id):
    p = program_id.lower()
    if any(x in p for x in ("bridge", "warp", "hyp_", "mailbox", "gmp", "connection", "vlink")):
        return "bridge/cross-chain"
    if any(x in p for x in ("stake", "staking", "delegator", "validator", "pondo", "whalepool")):
        return "staking/liquid staking"
    if any(x in p for x in ("token", "stablecoin", "usdc", "usad", "wrapped", "credits")):
        return "token/credit flow"
    if any(x in p for x in ("oracle", "checksum", "rate_limit")):
        return "oracle/governance guard"
    if any(x in p for x in ("nft", "ticket", "arcade", "mint")):
        return "NFT/game asset"
    if any(x in p for x in ("pool", "swap", "lend", "flash", "arcn", "dara")):
        return "DeFi pool/lending"
    return "application/other"


def write_reports(all_data, failures):
    report_dir = OUT / "reports"
    generated_at = datetime.now(timezone.utc).isoformat()
    lines = [
        "# Aleo context map",
        "",
        f"Generated at: {generated_at}",
        "",
        "Scope: active programs returned by Provable `/v2/{network}/metrics/programs`, with full source pulled from `/v2/{network}/program/{program_id}`.",
        "AleoScan registry pages were used as a secondary explorer reference, but local collection avoids Cloudflare-protected scraping.",
        "",
        "## Source folders",
        "",
        "- `raw/mainnet/program_sources/` and `raw/testnet/program_sources/`: downloaded Aleo Instruction sources.",
        "- `parsed/{network}/`: per-program JSON plus CSV summaries.",
        "- `graphs/dependencies.dot`: cross-program reference graph.",
        "- `reports/high_risk_async_paths.md`: async, delayed, conditional, and bridge/proof-sensitive paths.",
        "",
    ]
    for network, programs in all_data.items():
        total_calls = sum(p["calls_24h_or_metric_window"] for p in programs)
        lines.extend(
            [
                f"## {network}",
                "",
                f"- Programs collected: {len(programs)}",
                f"- Metric-window calls represented: {total_calls}",
                f"- Download failures: {len(failures.get(network, []))}",
                "",
                "| Program | Calls | Class | Functions | Mappings | Cross refs | Async/conditional paths |",
                "|---|---:|---|---:|---:|---|---:|",
            ]
        )
        for p in sorted(programs, key=lambda x: x["calls_24h_or_metric_window"], reverse=True):
            refs = ", ".join(p["cross_contract_refs"][:6])
            lines.append(
                f"| `{p['program_id']}` | {p['calls_24h_or_metric_window']} | {classify_path(p['program_id'])} | "
                f"{len(p['functions'])} | {len(p['mappings'])} | {refs or '-'} | {len(p['async_or_conditional_paths'])} |"
            )
        lines.append("")
    write_text(report_dir / "context_map.md", "\n".join(lines))

    risky = [
        "# Async, delayed, conditional, bridge/proof-sensitive paths",
        "",
        "Selection criteria: `async`, `finalize`, `.future`, `block.height`, branches/assertions/ternaries, `snark.verify`, or names indicating bridge/oracle/staking/token economic paths.",
        "",
    ]
    for network, programs in all_data.items():
        risky.append(f"## {network}")
        for p in sorted(programs, key=lambda x: x["calls_24h_or_metric_window"], reverse=True):
            important_name = classify_path(p["program_id"]) != "application/other"
            selected = p["async_or_conditional_paths"] or important_name
            if not selected:
                continue
            risky.append(f"### `{p['program_id']}` ({p['calls_24h_or_metric_window']} calls, {classify_path(p['program_id'])})")
            if p["cross_contract_refs"]:
                risky.append("Cross-contract refs: " + ", ".join(f"`{x}`" for x in p["cross_contract_refs"]))
            if p["mappings"]:
                risky.append("State mappings: " + ", ".join(f"`{x}`" for x in p["mappings"].keys()))
            for fn in p["async_or_conditional_paths"]:
                fndata = p["functions"][fn]
                flags = []
                if fndata["async"]:
                    flags.append("async")
                if fndata["has_finalize"]:
                    flags.append("finalize")
                if fndata["finalize_uses_block_height"]:
                    flags.append("block.height delay")
                if fndata["snark_verify"]:
                    flags.append("snark.verify")
                if fndata["cross_contract_refs"]:
                    flags.append("cross-contract")
                risky.append(f"- `{fn}`: {', '.join(flags) or 'conditional'}")
            risky.append("")
    write_text(report_dir / "high_risk_async_paths.md", "\n".join(risky))

    econ = [
        "# Economic paths and activity",
        "",
        "Calls are from Provable program metrics at collection time. TVL is not recomputed from private records; use this as a routing/activity map, not a balance attestation.",
        "",
    ]
    for network, programs in all_data.items():
        econ.append(f"## {network}")
        for p in sorted(programs, key=lambda x: x["calls_24h_or_metric_window"], reverse=True):
            cls = classify_path(p["program_id"])
            if cls == "application/other" and p["calls_24h_or_metric_window"] < 50:
                continue
            econ.append(
                f"- `{p['program_id']}`: {p['calls_24h_or_metric_window']} calls; class `{cls}`; "
                f"state `{', '.join(list(p['mappings'].keys())[:8]) or 'no public mappings parsed'}`."
            )
        econ.append("")
    write_text(report_dir / "economic_paths.md", "\n".join(econ))


def main():
    OUT.mkdir(parents=True, exist_ok=True)
    all_data = {}
    failures = {}
    for network in NETWORKS:
        programs, failed = collect_network(network)
        all_data[network] = programs
        failures[network] = failed
    write_graphs(all_data)
    write_reports(all_data, failures)
    write_text(
        OUT / "sources.json",
        json.dumps(
            {
                "provable_v2_program_metrics": f"{API}/{{network}}/metrics/programs",
                "provable_v2_program_source": f"{API}/{{network}}/program/{{program_id}}",
                "provable_v2_program_mappings": f"{API}/{{network}}/program/{{program_id}}/mappings",
                "provable_v1_latest_height": f"{API_V1}/{{network}}/latest/height",
                "aleo_docs_programs": "https://developer.aleo.org/apis/v2/programs/",
                "aleo_docs_program_metrics": "https://developer.aleo.org/apis/v2/program-metrics/",
                "aleo_docs_defi": "https://developer.aleo.org/apis/v2/de-fi/",
                "aleo_docs_staking": "https://developer.aleo.org/apis/v2/staking/",
                "aleoscan_registry_reference": "https://aleoscan.io/programs and https://testnet.aleoscan.io/programs",
            },
            indent=2,
        ),
    )
    print(f"Wrote Aleo context to {OUT}")


if __name__ == "__main__":
    main()
